File: frontend/src/lib/strategy-templates.ts
import strategyCatalog from "./strategy-catalog.json";

import type {
  LaunchStrategyConfig,
  OrganizeMethod,
  SessionStrategySelection,
  SessionStrategySummary,
  StrategyCautionLevel,
  StrategyDensity,
  StrategyLanguage,
  StrategyPrefixStyle,
  TaskType,
  StrategyTemplateId,
} from "@/types/session";

type DirectoryCatalogEntry = {
  id: string;
  labels: Record<StrategyLanguage, string>;
  categoryLabels: Record<StrategyLanguage, string>;
};

export interface StrategyTemplateMeta {
  id: StrategyTemplateId;
  label: string;
  description: string;
  applicableScenarios: string;
  directorySets: Record<StrategyDensity, DirectoryCatalogEntry[]>;
  defaultLanguage?: StrategyLanguage;
  defaultDensity?: StrategyDensity;
  defaultPrefixStyle?: StrategyPrefixStyle;
  defaultCautionLevel?: StrategyCautionLevel;
}

export interface StrategyOptionMeta<T extends string> {
  id: T;
  label: string;
  description: string;
}

type StrategyCatalog = {
  defaults: {
    templateId: StrategyTemplateId;
    language: StrategyLanguage;
    density: StrategyDensity;
    prefixStyle: StrategyPrefixStyle;
    cautionLevel: StrategyCautionLevel;
  };
  strategyTemplates: StrategyTemplateMeta[];
  languages: Array<StrategyOptionMeta<StrategyLanguage>>;
  densities: Array<StrategyOptionMeta<StrategyDensity>>;
  prefixStyles: Array<StrategyOptionMeta<StrategyPrefixStyle>>;
  cautionLevels: Array<StrategyOptionMeta<StrategyCautionLevel>>;
};

const catalog = strategyCatalog as StrategyCatalog;

export const STRATEGY_TEMPLATES: StrategyTemplateMeta[] = catalog.strategyTemplates;
export const LANGUAGE_OPTIONS: StrategyOptionMeta<StrategyLanguage>[] = catalog.languages;
export const DENSITY_OPTIONS: StrategyOptionMeta<StrategyDensity>[] = catalog.densities;
export const PREFIX_STYLE_OPTIONS: StrategyOptionMeta<StrategyPrefixStyle>[] = catalog.prefixStyles;
export const CAUTION_LEVEL_OPTIONS: StrategyOptionMeta<StrategyCautionLevel>[] = catalog.cautionLevels;

export function taskTypeForOrganizeMode(organizeMode: SessionStrategySelection["organize_mode"] | null | undefined): TaskType {
  return organizeMode === "incremental" ? "organize_into_existing" : "organize_full_directory";
}

export function organizeMethodForOrganizeMode(
  organizeMode: SessionStrategySelection["organize_mode"] | null | undefined,
): OrganizeMethod {
  return organizeMode === "incremental"
    ? "assign_into_existing_categories"
    : "categorize_into_new_structure";
}

export function taskTypeLabel(taskType: TaskType): string {
  return taskType === "organize_into_existing" ? "归入已有目录" : "整理整个目录";
}

export const DEFAULT_STRATEGY_SELECTION: SessionStrategySelection = {
  template_id: catalog.defaults.templateId,
  organize_mode: "initial",
  task_type: "organize_full_directory",
  organize_method: "categorize_into_new_structure",
  destination_index_depth: 2,
  language: catalog.defaults.language,
  density: catalog.defaults.density,
  prefix_style: catalog.defaults.prefixStyle,
  caution_level: catalog.defaults.cautionLevel,
  note: "",
};

export function getTemplateMeta(templateId: StrategyTemplateId): StrategyTemplateMeta {
  return STRATEGY_TEMPLATES.find((template) => template.id === templateId) || STRATEGY_TEMPLATES[0];
}

function buildDirectoryName(
  entry: DirectoryCatalogEntry,
  index: number,
  language: StrategyLanguage,
  prefixStyle: StrategyPrefixStyle,
): string {
  const baseLabel = entry.labels[language] || entry.labels.zh;
  if (!baseLabel) {
    return "";
  }
  if (prefixStyle === "numeric") {
    return `${String(index + 1).padStart(2, "0")}_${baseLabel}`;
  }
  if (prefixStyle === "category") {
    const categoryLabel = entry.categoryLabels[language] || entry.categoryLabels.zh || baseLabel;
    return `[${categoryLabel}] ${baseLabel}`;
  }
  return baseLabel;
}

export function buildPreviewDirectories(strategy: Pick<SessionStrategySelection, "template_id" | "language" | "density" | "prefix_style">): string[] {
  const template = getTemplateMeta(strategy.template_id);
  const entries = template.directorySets[strategy.density] || template.directorySets.normal || [];
  return entries
    .map((entry, index) => buildDirectoryName(entry, index, strategy.language, strategy.prefix_style))
    .filter(Boolean);
}

export function getSuggestedSelection(
  templateId: StrategyTemplateId,
): Pick<SessionStrategySelection, "language" | "density" | "prefix_style" | "caution_level"> {
  const template = getTemplateMeta(templateId);
  return {
    language: template.defaultLanguage || DEFAULT_STRATEGY_SELECTION.language,
    density: template.defaultDensity || DEFAULT_STRATEGY_SELECTION.density,
    prefix_style: template.defaultPrefixStyle || DEFAULT_STRATEGY_SELECTION.prefix_style,
    caution_level: template.defaultCautionLevel || DEFAULT_STRATEGY_SELECTION.caution_level,
  };
}

function isValidTemplateId(value: unknown): value is StrategyTemplateId {
  return STRATEGY_TEMPLATES.some((template) => template.id === value);
}

function isValidLanguage(value: unknown): value is StrategyLanguage {
  return LANGUAGE_OPTIONS.some((option) => option.id === value);
}

function isValidDensity(value: unknown): value is StrategyDensity {
  return DENSITY_OPTIONS.some((option) => option.id === value);
}

function isValidPrefixStyle(value: unknown): value is StrategyPrefixStyle {
  return PREFIX_STYLE_OPTIONS.some((option) => option.id === value);
}

function isValidCautionLevel(value: unknown): value is StrategyCautionLevel {
  return CAUTION_LEVEL_OPTIONS.some((option) => option.id === value);
}

function isValidOrganizeMethod(value: unknown): value is OrganizeMethod {
  return value === "categorize_into_new_structure" || value === "assign_into_existing_categories";
}

export function getLaunchStrategyFromConfig(config?: LaunchStrategyConfig | null): SessionStrategySelection {
  const templateId = isValidTemplateId(config?.LAUNCH_DEFAULT_TEMPLATE_ID)
    ? config.LAUNCH_DEFAULT_TEMPLATE_ID
    : DEFAULT_STRATEGY_SELECTION.template_id;
  const suggested = getSuggestedSelection(templateId);
  const organizeMethod = isValidOrganizeMethod(config?.LAUNCH_DEFAULT_ORGANIZE_METHOD)
    ? config.LAUNCH_DEFAULT_ORGANIZE_METHOD
    : DEFAULT_STRATEGY_SELECTION.organize_method;
  const organizeMode = organizeMethod === "assign_into_existing_categories" ? "incremental" : "initial";

  return {
    template_id: templateId,
    organize_mode: organizeMode,
    task_type: taskTypeForOrganizeMode(organizeMode),
    organize_method: organizeMethod,
    destination_index_depth: 2,
    language: isValidLanguage(config?.LAUNCH_DEFAULT_LANGUAGE)
      ? config.LAUNCH_DEFAULT_LANGUAGE
      : suggested.language,
    density: isValidDensity(config?.LAUNCH_DEFAULT_DENSITY)
      ? config.LAUNCH_DEFAULT_DENSITY
      : suggested.density,
    prefix_style: isValidPrefixStyle(config?.LAUNCH_DEFAULT_PREFIX_STYLE)
      ? config.LAUNCH_DEFAULT_PREFIX_STYLE
      : suggested.prefix_style,
    caution_level: isValidCautionLevel(config?.LAUNCH_DEFAULT_CAUTION_LEVEL)
      ? config.LAUNCH_DEFAULT_CAUTION_LEVEL
      : suggested.caution_level,
    target_profile_id: typeof config?.LAUNCH_DEFAULT_TARGET_PROFILE_ID === "string"
      ? config.LAUNCH_DEFAULT_TARGET_PROFILE_ID
      : undefined,
    note: typeof config?.LAUNCH_DEFAULT_NOTE === "string" ? config.LAUNCH_DEFAULT_NOTE : "",
  };
}

export function shouldSkipLaunchStrategyPrompt(config?: LaunchStrategyConfig | null): boolean {
  return Boolean(config?.LAUNCH_SKIP_STRATEGY_PROMPT);
}

export function buildStrategySummary(strategy: SessionStrategySelection): SessionStrategySummary {
  const template = getTemplateMeta(strategy.template_id);
  const taskType = strategy.task_type || taskTypeForOrganizeMode(strategy.organize_mode);
  const organizeMethod = strategy.organize_method || organizeMethodForOrganizeMode(strategy.organize_mode);
  const languageLabel = LANGUAGE_OPTIONS.find((item) => item.id === strategy.language)?.label || "中文目录";
  const densityLabel = DENSITY_OPTIONS.find((item) => item.id === strategy.density)?.label || "常规分类";
  const prefixStyleLabel = PREFIX_STYLE_OPTIONS.find((item) => item.id === strategy.prefix_style)?.label || "无前缀";
  const cautionLabel = CAUTION_LEVEL_OPTIONS.find((item) => item.id === strategy.caution_level)?.label || "平衡";

  return {
    ...strategy,
    task_type: taskType,
    organize_method: organizeMethod,
    task_type_label: taskTypeLabel(taskType),
    template_label: template.label,
    template_description: template.description,
    organize_mode_label: taskTypeLabel(taskType),
    language_label: languageLabel,
    density_label: densityLabel,
    prefix_style_label: prefixStyleLabel,
    caution_level_label: cautionLabel,
    preview_directories: buildPreviewDirectories(strategy),
  };
}

File: file_pilot/organize/strategy_templates.py
from __future__ import annotations

import json
from copy import deepcopy
from functools import lru_cache
from pathlib import Path


_CATALOG_PATH = Path(__file__).resolve().parents[2] / "frontend" / "src" / "lib" / "strategy-catalog.json"


@lru_cache(maxsize=1)
def _load_strategy_catalog() -> dict:
    payload = json.loads(_CATALOG_PATH.read_text(encoding="utf-8"))

    defaults = payload.get("defaults", {})
    strategy_templates = {
        item["id"]: {
            "id": item["id"],
            "label": item["label"],
            "description": item["description"],
            "applicable_scenarios": item.get("applicableScenarios", ""),
            "default_language": item.get("defaultLanguage"),
            "default_density": item.get("defaultDensity"),
            "default_prefix_style": item.get("defaultPrefixStyle"),
            "default_caution_level": item.get("defaultCautionLevel"),
            "directory_sets": deepcopy(item.get("directorySets", {})),
            "prompt_fragment": item.get("promptFragment", ""),
        }
        for item in payload.get("strategyTemplates", [])
    }
    languages = {
        item["id"]: {
            "id": item["id"],
            "label": item["label"],
            "description": item.get("description", ""),
            "prompt_fragment": item.get("promptFragment", ""),
        }
        for item in payload.get("languages", [])
    }
    densities = {
        item["id"]: {
            "id": item["id"],
            "label": item["label"],
            "description": item.get("description", ""),
            "prompt_fragment": item.get("promptFragment", ""),
        }
        for item in payload.get("densities", [])
    }
    prefix_styles = {
        item["id"]: {
            "id": item["id"],
            "label": item["label"],
            "description": item.get("description", ""),
            "prompt_fragment": item.get("promptFragment", ""),
        }
        for item in payload.get("prefixStyles", [])
    }
    caution_levels = {
        item["id"]: {
            "id": item["id"],
            "label": item["label"],
            "description": item.get("description", ""),
            "prompt_fragment": item.get("promptFragment", ""),
        }
        for item in payload.get("cautionLevels", [])
    }
    return {
        "defaults": defaults,
        "strategy_templates": strategy_templates,
        "languages": languages,
        "densities": densities,
        "prefix_styles": prefix_styles,
        "caution_levels": caution_levels,
    }


_CATALOG = _load_strategy_catalog()

DEFAULT_TEMPLATE_ID = str(_CATALOG["defaults"].get("templateId", "general_downloads") or "general_downloads")
DEFAULT_LANGUAGE = str(_CATALOG["defaults"].get("language", "zh") or "zh")
DEFAULT_DENSITY = str(_CATALOG["defaults"].get("density", "normal") or "normal")
DEFAULT_PREFIX_STYLE = str(_CATALOG["defaults"].get("prefixStyle", "none") or "none")
DEFAULT_CAUTION_LEVEL = str(_CATALOG["defaults"].get("cautionLevel", "balanced") or "balanced")
DEFAULT_ORGANIZE_MODE = "initial"
DEFAULT_DESTINATION_INDEX_DEPTH = 2
TASK_TYPE_ORGANIZE_FULL_DIRECTORY = "organize_full_directory"
TASK_TYPE_ORGANIZE_INTO_EXISTING = "organize_into_existing"
DEFAULT_TASK_TYPE = TASK_TYPE_ORGANIZE_FULL_DIRECTORY
ORGANIZE_METHOD_CATEGORIZE_INTO_NEW_STRUCTURE = "categorize_into_new_structure"
ORGANIZE_METHOD_ASSIGN_INTO_EXISTING_CATEGORIES = "assign_into_existing_categories"
DEFAULT_ORGANIZE_METHOD = ORGANIZE_METHOD_CATEGORIZE_INTO_NEW_STRUCTURE

STRATEGY_TEMPLATES = deepcopy(_CATALOG["strategy_templates"])
LANGUAGES = deepcopy(_CATALOG["languages"])
DENSITIES = deepcopy(_CATALOG["densities"])
PREFIX_STYLES = deepcopy(_CATALOG["prefix_styles"])
CAUTION_LEVELS = deepcopy(_CATALOG["caution_levels"])


def task_type_for_organize_mode(organize_mode: str | None) -> str:
    normalized = str(organize_mode or DEFAULT_ORGANIZE_MODE).strip().lower()
    if normalized == "incremental":
        return TASK_TYPE_ORGANIZE_INTO_EXISTING
    return TASK_TYPE_ORGANIZE_FULL_DIRECTORY


def organize_method_for_task_type(task_type: str | None) -> str:
    normalized = str(task_type or DEFAULT_TASK_TYPE).strip().lower()
    if normalized == TASK_TYPE_ORGANIZE_INTO_EXISTING:
        return ORGANIZE_METHOD_ASSIGN_INTO_EXISTING_CATEGORIES
    return ORGANIZE_METHOD_CATEGORIZE_INTO_NEW_STRUCTURE


def organize_method_for_organize_mode(organize_mode: str | None) -> str:
    return organize_method_for_task_type(task_type_for_organize_mode(organize_mode))


def task_type_for_organize_method(organize_method: str | None) -> str:
    normalized = str(organize_method or DEFAULT_ORGANIZE_METHOD).strip().lower()
    if normalized == ORGANIZE_METHOD_ASSIGN_INTO_EXISTING_CATEGORIES:
        return TASK_TYPE_ORGANIZE_INTO_EXISTING
    return TASK_TYPE_ORGANIZE_FULL_DIRECTORY


def organize_mode_for_organize_method(organize_method: str | None) -> str:
    return organize_mode_for_task_type(task_type_for_organize_method(organize_method))


def organize_mode_for_task_type(task_type: str | None) -> str:
    normalized = str(task_type or DEFAULT_TASK_TYPE).strip().lower()
    if normalized == TASK_TYPE_ORGANIZE_INTO_EXISTING:
        return "incremental"
    return "initial"


def task_type_label(task_type: str | None) -> str:
    normalized = str(task_type or DEFAULT_TASK_TYPE).strip().lower()
    if normalized == TASK_TYPE_ORGANIZE_INTO_EXISTING:
        return "归入已有目录"
    return "整理整个目录"


def _normalize_task_type(value: str | None) -> str:
    normalized = str(value or "").strip().lower()
    if normalized in {TASK_TYPE_ORGANIZE_FULL_DIRECTORY, TASK_TYPE_ORGANIZE_INTO_EXISTING}:
        return normalized
    return ""


def _normalize_organize_method(value: str | None) -> str:
    normalized = str(value or "").strip().lower()
    if normalized in {
        ORGANIZE_METHOD_CATEGORIZE_INTO_NEW_STRUCTURE,
        ORGANIZE_METHOD_ASSIGN_INTO_EXISTING_CATEGORIES,
    }:
        return normalized
    return ""


def _directory_entries_for(template: dict, density: str) -> list[dict]:
    directory_sets = template.get("directory_sets", {})
    return list(directory_sets.get(density) or directory_sets.get(DEFAULT_DENSITY) or [])


def _format_directory_name(entry: dict, *, index: int, language: str, prefix_style: str) -> str:
    labels = dict(entry.get("labels", {}))
    category_labels = dict(entry.get("categoryLabels", {}))
    base_label = str(labels.get(language) or labels.get(DEFAULT_LANGUAGE) or "").strip()
    if not base_label:
        return ""
    if prefix_style == "numeric":
        return f"{index + 1:02d}_{base_label}"
    if prefix_style == "category":
        category_label = str(category_labels.get(language) or category_labels.get(DEFAULT_LANGUAGE) or base_label).strip()
        return f"[{category_label}] {base_label}"
    return base_label


def build_preview_directories(template_id: str, *, language: str, density: str, prefix_style: str) -> list[str]:
    template = STRATEGY_TEMPLATES.get(template_id)
    if not template:
        return []

    directories: list[str] = []
    for index, entry in enumerate(_directory_entries_for(template, density)):
        directory_name = _format_directory_name(entry, index=index, language=language, prefix_style=prefix_style)
        if directory_name:
            directories.append(directory_name)
    return directories


def list_strategy_templates() -> list[dict]:
    return [deepcopy(template) for template in STRATEGY_TEMPLATES.values()]


def normalize_strategy_selection(raw: dict | None = None) -> dict:
    payload = raw or {}
    template_id = payload.get("template_id")
    if template_id not in STRATEGY_TEMPLATES:
        template_id = DEFAULT_TEMPLATE_ID

    template = STRATEGY_TEMPLATES[template_id]

    language = payload.get("language") or template.get("default_language") or DEFAULT_LANGUAGE
    if language not in LANGUAGES:
        language = template.get("default_language") or DEFAULT_LANGUAGE

    density = payload.get("density") or template.get("default_density") or DEFAULT_DENSITY
    if density not in DENSITIES:
        density = template.get("default_density") or DEFAULT_DENSITY

    prefix_style = payload.get("prefix_style") or template.get("default_prefix_style") or DEFAULT_PREFIX_STYLE
    if prefix_style not in PREFIX_STYLES:
        prefix_style = template.get("default_prefix_style") or DEFAULT_PREFIX_STYLE

    caution_level = payload.get("caution_level") or template.get("default_caution_level") or DEFAULT_CAUTION_LEVEL
    if caution_level not in CAUTION_LEVELS:
        caution_level = template.get("default_caution_level") or DEFAULT_CAUTION_LEVEL

    requested_task_type = _normalize_task_type(payload.get("task_type"))
    requested_organize_mode = str(payload.get("organize_mode") or "").strip().lower()
    requested_organize_method = _normalize_organize_method(payload.get("organize_method"))
    if requested_organize_mode not in {"initial", "incremental"}:
        requested_organize_mode = ""
    if requested_task_type and requested_organize_mode:
        expected_mode = organize_mode_for_task_type(requested_task_type)
        if expected_mode != requested_organize_mode:
            raise ValueError("TASK_TYPE_CONFLICT")
    if requested_task_type and requested_organize_method:
        expected_method = organize_method_for_task_type(requested_task_type)
        if expected_method != requested_organize_method:
            raise ValueError("TASK_TYPE_CONFLICT")
    if requested_organize_mode and requested_organize_method:
        expected_method = organize_method_for_organize_mode(requested_organize_mode)
        if expected_method != requested_organize_method:
            raise ValueError("TASK_TYPE_CONFLICT")
    organize_method = (
        requested_organize_method
        or (organize_method_for_task_type(requested_task_type) if requested_task_type else "")
        or (organize_method_for_organize_mode(requested_organize_mode) if requested_organize_mode else "")
        or DEFAULT_ORGANIZE_METHOD
    )
    task_type = requested_task_type or task_type_for_organize_method(organize_method)
    organize_mode = requested_organize_mode or organize_mode_for_organize_method(organize_method)

    try:
        destination_index_depth = int(payload.get("destination_index_depth") or DEFAULT_DESTINATION_INDEX_DEPTH)
    except (TypeError, ValueError):
        destination_index_depth = DEFAULT_DESTINATION_INDEX_DEPTH
    destination_index_depth = max(1, min(3, destination_index_depth))

    note = str(payload.get("note") or "").strip()
    output_dir = str(payload.get("output_dir") or "").strip()
    target_profile_id = str(payload.get("target_profile_id") or "").strip()
    new_directory_root = str(payload.get("new_directory_root") or "").strip()
    review_root = str(payload.get("review_root") or "").strip()
    target_directories = [
        str(item).strip()
        for item in (payload.get("target_directories") or [])
        if str(item).strip()
    ]
    preview_directories = build_preview_directories(
        template_id,
        language=language,
        density=density,
        prefix_style=prefix_style,
    )

    return {
        "template_id": template_id,
        "template_label": template["label"],
        "template_description": template["description"],
        "language": language,
        "language_label": LANGUAGES[language]["label"],
        "density": density,
        "density_label": DENSITIES[density]["label"],
        "prefix_style": prefix_style,
        "prefix_style_label": PREFIX_STYLES[prefix_style]["label"],
        "caution_level": caution_level,
        "caution_level_label": CAUTION_LEVELS[caution_level]["label"],
        "task_type": task_type,
        "task_type_label": task_type_label(task_type),
        "organize_method": organize_method,
        "organize_mode": organize_mode,
        "organize_mode_label": task_type_label(task_type),
        "destination_index_depth": destination_index_depth,
        "output_dir": output_dir,
        "target_profile_id": target_profile_id,
        "target_directories": target_directories,
        "new_directory_root": new_directory_root,
        "review_root": review_root,
        "note": note,
        "preview_directories": preview_directories,
    }


def build_strategy_prompt_fragment(selection: dict | None = None) -> str:
    normalized = normalize_strategy_selection(selection)
    template = STRATEGY_TEMPLATES[normalized["template_id"]]
    caution = CAUTION_LEVELS[normalized["caution_level"]]
    if normalized["organize_mode"] == "initial":
        language = LANGUAGES[normalized["language"]]
        density = DENSITIES[normalized["density"]]
        prefix_style = PREFIX_STYLES[normalized["prefix_style"]]
        lines = [
            "当前固定整理策略（必须优先遵守）：",
            f"- 主模板：{normalized['template_label']}。{template['description']}",
            template["prompt_fragment"],
            language["prompt_fragment"],
            density["prompt_fragment"],
            prefix_style["prompt_fragment"],
            caution["prompt_fragment"],
            "当前任务类型：整理整个目录。只处理当前规划范围内的条目，不扩展到范围外；可以创建新的目标目录。",
        ]
    else:
        lines = [
            "当前固定整理策略（必须优先遵守）：",
            "- 当前任务类型：归入已有目录。只能使用用户显式配置的目标目录；拿不准的条目交给系统放入待确认区，不要重新设计分类体系。",
            "- 父目录不会自动授权子目录；需要子目录作为去向时，用户必须把该子目录单独加入目标目录配置。",
            caution["prompt_fragment"],
        ]
    if normalized["new_directory_root"]:
        lines.append(f"- 新目录生成位置：{normalized['new_directory_root']}")
    if normalized["review_root"]:
        lines.append(f"- 待确认区目录位置：{normalized['review_root']}")
    if normalized["note"]:
        lines.append(f"用户补充说明：{normalized['note']}")
    return "\n".join(lines)
